src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -970,13 +1371,194 @@ fn slice_pat_covered_by_constructor<'tcx>(
+                    if let Some(r) = IntRange::from_pat(tcx, row[0]) {
+                        // We're only interested in endpoints that lie (at least partially)
+                        // within the subrange domain.
+                        if let Some(r) = ctor_range.intersection(&r) {
+                            let (r_lo, r_hi) = r.range.into_inner();
+                            add_endpoints(&mut points, r_lo, r_hi);
+                        }
+                    }
+                }
+
+                // The patterns were iterated in an arbitrary order (i.e. in the order the user
+                // wrote them), so we need to make sure our endpoints are sorted.
+                let mut points: Vec<(u128, Endpoint)> = points.into_iter().collect();
+                points.sort_unstable_by_key(|(x, _)| *x);
+                let mut points = points.into_iter();
+                let mut a = points.next().unwrap();
+
+                // Iterate through pairs of points, adding the subranges to `split_ctors`.
+                // We have to be careful about the orientation of the points as endpoints, to make
+                // sure we're enumerating precisely the correct ranges. Too few and the matching is
+                // actually incorrect. Too many and our diagnostics are poorer. This involves some
+                // case analysis.
+                while let Some(b) = points.next() {
+                    // a < b (strictly)
+                    if let Endpoint::Both = a.1 {
+                        split_ctors.push(IntRange::range_to_ctor(tcx, ty, a.0..=a.0));
+                    }
+                    let c = match a.1 {
+                        Endpoint::Start => a.0,
+                        Endpoint::End | Endpoint::Both => a.0 + 1,
+                    };
+                    let d = match b.1 {
+                        Endpoint::Start | Endpoint::Both => b.0 - 1,
+                        Endpoint::End => b.0,
+                    };
+                    // In some cases, we won't need an intermediate range between two ranges
+                    // lie immediately adjacent to one another.
+                    if c <= d {
+                        split_ctors.push(IntRange::range_to_ctor(tcx, ty, c..=d));
+                    }
+
+                    a = b;
+                }
+            }
+            // Any other constructor can be used unchanged.
+            _ => split_ctors.push(ctor),
+        }
+    }
+
+    split_ctors
+}
+
+/// Check whether there exists any shared value in either `ctor` or `pat` by intersecting them.
+fn constructor_intersects_pattern<'p, 'a: 'p, 'tcx: 'a>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    ctor: &Constructor<'tcx>,
+    pat: &'p Pattern<'tcx>,
+) -> Option<Vec<&'p Pattern<'tcx>>> {
+    if should_treat_range_exhaustively(tcx, ctor) {
+        match (IntRange::from_ctor(tcx, ctor), IntRange::from_pat(tcx, pat)) {
+            (Some(ctor), Some(pat)) => ctor.intersection(&pat).map(|_| vec![]),
```

Comment:
Add an assertion here that the constructor does not partially overlap the pattern (i.e. either that `ctor` is a proper subset of `pat` or that it is disjoint from it)?

`split_grouped_constructors` should guarantee that (I think that's the correct guarantee) and it will prevent surprises.